### ISSUE TITLE
Angularjs UI-Router Issue

### ISSUE THREAD

**Issue by ronster37** [qzind/tray#36]
The stringify tools method is causing issues with angularjs's ui-router as it is forcing the toJSON index on the prototype.

As a workaround I have added this code to the stringify function:

```diff
            stringify: function(object) {
                //old versions of prototype affect stringify
                var pjson = Array.prototype.toJSON;
                delete Array.prototype.toJSON;

                var result = JSON.stringify(object);

+                if (pjson) {
                    Array.prototype.toJSON = pjson;
+                }

                return result;
            },
```

**Comment by tresf** [comment]
Changed to `diff` format for better readability.  Brett, we should have done this to begin with I think, if not, as @ronster37 states, frameworks may inadvertently go against the prototype.

**Comment by bberenz** [comment]
I've no problem making this change, but I want to understand it further first in case there are other problems around this function. As in my understanding, if `Array.prototype.toJSON` is `undefined`, setting it to `undefined` shouldn't have an effect on the prototype.

@ronster37 I've not used ui-router before to have a thorough understanding of the problem created here, and my basic setup with it didn't seem to reproduce this. Can you elaborate on how this call is affecting your setup?

**Comment by tresf** [comment]
@bberenz I think the scenario is namespace.  You're assuming `Array.prototype.toJSON` is equal to `null` an we're leaving it alone, but in reality we're assigning it, adding it as a namespace and that can have a different effect on a library that searches for it over it not existing at all.

**Comment by bberenz** [comment]
Not sure I'm following. I know assigning `undefined` is different from the value not existing. But accessing either comes back with the same result, so if something searches for it, it will look the same.

**Comment by tresf** [comment]
@bberenz here you are... 

```javascript
var obj = new Array();

var foo = Array.prototype.foo;

alert('foo' in Array.prototype);

delete Array.prototype.foo;

alert('foo' in Array.prototype);

Array.prototype.foo = foo;

alert('foo' in Array.prototype);
```

**Comment by bberenz** [comment]
Right, like I said:
>The only thing they differ on is that explicitly setting undefined keeps the key around.

But yet again, my question is how is this affecting ui-router? As this particular condition only affects if they are looping on keys _without_ the `hasOwnProperty` check, which explicitly looks for - and excludes - values that came from parents (such as the prototype). `toJSON` won't show up as an ownProperty on any array with this check in place, meaning something might've been coded poorly.

**Comment by tresf** [comment]
> But yet again, my question is how is this affecting ui-router

Well, if you can't reproduce, they've probably fixed it.  @ronster37's version of ui-router may be relevant here as well.

I tried comparing [`legacy`](https://github.com/angular-ui/ui-router/tree/legacy) ui-router versus [`master`](https://github.com/angular-ui/ui-router/tree/master), but the API was a bit hard to follow through the web browser.  They use a function called `toJson` internally, but even the definition of that in ui-router has changed considerably between versions, and the two branches each seem to support various AngularJS versions (1.0.5 - 1.5.0) vs (1.2.0 - 1.5.0).

So I suppose it depends how widespread the underlying bugged logic is.  My vote is to just fix it.

**Comment by ronster37** [comment]
Hey guys just wanted to update that they fixed it on version 0.2.16, I was on 0.2.15 of ui-router. But in any case not sure if this would be a problem with other frameworks, this fix could just be for insurance.

Thanks for the help guys I appreciate it :)

**Comment by tresf** [comment]
> Hey guys just wanted to update that they fixed it on version 0.2.16, I was on 0.2.15 of ui-router. But in any case not sure if this would be a problem with other frameworks, this fix could just be for insurance.
>
> Thanks for the help guys I appreciate it :)

Terrific.  That will be our recommendation if it surfaces again.  Marking as `wontfix` for now, can revisit if it starts affecting more people. 👍
